Projectile mod element and Remove ranged item mod element

Goldorion opened the pull request:
This PR is the long-awaited PR that adds a Projectile mod element and splits ranged items into 2 mod elements.

### Important notes:
- Each ranged item from an old workspace = 1 projectile mod element + 1 item mod element (converters included)
- The ranged item mod element has been entirely deleted and moved to other mod elements (We won't miss you 🫡)
- Custom projectiles are still based on the arrows
- Every reference to `ammo` and `bullet` has been changed to `projectile`, so we can uniformize and generalize all texts. Tooltips have also been changed (~50% of changed files). Russian tips (texts only) need to be changed by @Defeatomizer.
- `ammoItem` parameter in the ranged item ME has been changed to a projectile selector (if it was empty, the projectile is still selected)
- Some parts of the code have been changed/made to support future features that should come after this PR will be merged.
- Variables in both GUIs now have new names to represent what they are (e.g. the weird `sbbp` and `selp` panel names)
- The normal model of the ranged item mod element has been added inside the item mod element as `Ranged item`. When ranged options are enabled, MCreator will automatically select this new built-in model if `Normal` was used.

This is the workspace I used to test everything (including converters). The workspace has been made and exported using the latest MCreator 2022.3 version. Everything required should be included. [test_workspace.zip](https://github.com/MCreator/MCreator/files/10388454/test_workspace.zip)

KlemenDEV commented:
I think it would make sense to merge the ranged item to "projectile + regular item ME". Most of the properties of the ranged and regular item are shared.

We would just need a new advanced properties section or a new section such as for inventory where one can define ammo, projectile, and shooting settings.

Atten007 commented:
Very cool PR, hope, that this can be added into MCreator very soon!

Goldorion commented:
> I think it would make sense to merge the ranged item to "projectile + regular item ME". Most of the properties of the ranged and regular item are shared.
> 
> We would just need a new advanced properties section or a new section such as for inventory where one can define ammo, projectile, and shooting settings.

It would make the PR even bigger. Is it ok for you?

KlemenDEV commented:
> However, why should we merge the projectile element in the item element too? What would be the point? We would still have the same problems, without saying a projectile is an entity only.

Ranged item without the entity (which is now projectile ME) is more or less item. Also:

* food was merged to item and it was much more different than item compared to ranged item
* inventory item is also specific parameter of item but is not separate ME
* it would make porting to new versions easier if there is less duplication of stuff

Goldorion commented:
> > However, why should we merge the projectile element in the item element too? What would be the point? We would still have the same problems, without saying a projectile is an entity only.
> 
> Ranged item without the entity (which is now projectile ME) is more or less item. Also:
> 
>     * food was merged to item and it was much more different than item compared to ranged item
> 
>     * inventory item is also specific parameter of item but is not separate ME
> 
>     * it would make porting to new versions easier if there is less duplication of stuff

I didn't mean to never merge the ranged item with the item. I just wanted to say that doing this in this specific PR in addition to everything else that is already added, changed and removed would make it very big and long to review. It can easily be made in another PR just after this one without breaking workspaces. I totally agree with you on the fact they need to be merged. I was just questioning if doing it in this PR is the best idea.

KlemenDEV commented:
> What would be the point?

My reply was answer to this question you said.

And also to this:

> why should we merge the projectile element in the item element too?

NerdyPuzzle commented:
I think merging them all into the item element would be a bit messy and confusing

KlemenDEV commented:
> I think merging them all into the item element would be a bit messy and confusing

Why did we merge food then? And why is there not a separate mod element for an item with inventory?

NerdyPuzzle commented:
Having all of the model/texture selectors and triggers in the same element seems a little odd, but I wouldn't be against the change if it is deemed the best option.

KlemenDEV commented:
> Having all of the model/texture selectors and triggers in the same element seems a little odd, but I wouldn't be against the change if it is deemed the best option.

The projectile with all parameters is now its own mod element. Items already have texture and model selector, don't they?

There would be a new section, ranged actions, that would have those parameters (probably some more):

* Enable ranged actions for this item
* Projectile to shoot
* Ammo item
* Shooting power

This way one could also make eg. ranged item with inventory, otherwise this is not possible.

If enabled, then code would only be additionally generated in event triggers that are needed for the ranged item.

Benefits:

* fewer MEs (also following the spirit of merging food to item)
* easier to port to new MC versions
* less maintainability in the long run
* more features can apply to the ranged item (features otherwise limited to regular items)

Drawbacks:

* this PR is larger
* item templates will be more bloated

@AlsoSomeoneElse, what is your take on this?

NerdyPuzzle commented:
Oh, I thought you also wanted to merge projectiles into items, I misinterpreted the reply. I think if the merging it so be done, the option to add multiple textures for pulling stages could also be introduced as a toggleable option in that section?

KlemenDEV commented:
> the option to add multiple textures for pulling stages could also be introduced as a toggleable option in that section?

This is probably something the user would handle with item properties if desired once this feature is merged

NerdyPuzzle commented:
Right, I forgot that was planned.

Goldorion commented:
> Ammo item
Shooting power

Ammo item is the projectile selector (the ammo is inside the projectile) The shooting power is also declared in the projectile (the things we discussed yesterday in our general)

> Oh, I thought you also wanted to merge projectiles into items, I misinterpreted the reply.

This is also what I thought originally.

Goldorion commented:
marking it as a draft until I've finished removing the ranged item mod element.

Defeatomizer commented:
Just make sure it's a separate **Ranged actions** page, not addition to **Advanced properties** page (to follow food ME).

Goldorion commented:
> Just make sure it's a separate **Ranged actions** page, not addition to **Advanced properties** page (to follow food ME).

It's the first thing I tried and personally, I think it fits better on the advanced properties page.

KlemenDEV commented:
> It's the first thing I tried and personally, I think it fits better on the advanced properties page.

Could work I guess, but maybe make them left and right, not top and bottom

Defeatomizer commented on plugins/generator-1.18.2/forge-1.18.2/templates/item/item.java.ftl:
```diff
@@ -253,5 +253,84 @@ public class ${name}Item extends Item {
 	}
 	</#if>
+
+	<#if data.shootConstantly>
```
AFAIU, this part should only be generated if ranged actions are enabled.

Goldorion replied:
Don't start reviewing what I said was not finished. I literally tested nothing yet, except the UI.

Defeatomizer reviewed (commented):
Some translation keys are missing, e.g. `elementgui.item.event_on_use`.

Goldorion commented:
> Some translation keys are missing, e.g. `elementgui.item.event_on_use`.

I know, I wanted to do and finish everything in a single commit, but the converter causes me a lot of problems.

Defeatomizer commented:
You are free to continue working on the feature.

Atten007 commented:
@Goldorion Will this be added into MCreator 2023.1?

NerdyPuzzle commented:
> Goldorion Will this be added into MCreator 2023.1?

Please do not needlessly ping contributors. This page is only for discussion related to the development of these features.

Goldorion commented:
> @Goldorion Will this be added into MCreator 2023.1?

it's not finished, it's huge, it's breaking changes, and it's the third snapshot next, so absolutely not.

Defeatomizer commented:
Still the same error:
```
    /tmp/mcreator_test_workspace3249868212655281871/src/main/java/net/mcreator/test_mod/item/Exampleitem7Item.java:150: error: method releaseUsing(ItemStack,Level,LivingEntity,int) is already defined in class Exampleitem7Item
    	public void releaseUsing(ItemStack itemstack, Level world, LivingEntity entityLiving, int timeLeft) {
    	            ^

Defeatomizer commented on plugins/mcreator-localization/help/default/item/enable_ranged_item.md:
```diff
@@ -0,0 +1 @@
+By enabling this parameter, the item will become a ranged item allowing it to shoot a projectile, like bows and crossbows.
```
Установите этот флажок, чтобы предмет можно было использовать для дальнего боя подобно луку, арбалету, снежку или огненному шару.
```suggestion
If you check this box, the item will become a ranged item allowing it to shoot a projectile, like bows and crossbows.
```

Defeatomizer commented on plugins/mcreator-localization/help/default/item/projectile.md:
```diff
@@ -0,0 +1,2 @@
+This is the projectile type this ranged item can shoot.
```
Определяет, какими снарядами можно стрелять из этого предмета.
Можно использовать один из "ванильных" снарядов или создать собственный.

Defeatomizer reviewed (commented):
Help tips translations.

Goldorion commented:
I finished moving everything related to the ranged item inside the item mod element. I (re)tested everything and it should work correctly. There are only Russian tips that still need to be changed, but I can't do them myself 🤷

Defeatomizer commented:
`item/projectile` and `rangeditem/projectile` help tips explain the same thing, which one should be used?
Also, first half of `projectile/projectile_item` is written in French, but has more information than a help tip with the same name in French translations directory, so the same question: what text should be used?

KlemenDEV commented on plugins/mcreator-localization/help/default/rangeditem/projectile.md:
```diff
@@ -0,0 +1,4 @@
+This is the projectile that is used by this ranged item.
```
This file should probably now be moved to the item folder

Defeatomizer replied:
I already asked Goldorion about which of these files should be kept (check my most recent comment).

Defeatomizer commented:
@Goldorion, I updated help tips for you, the only thing I did not change is the bottom line of French translation of `item/projectile`.

Goldorion commented:
> Another thing I noticed: the converter should probably set item's animation to `"bow"` if it's not specified in its JSON definition.

Good idea

Defeatomizer commented:
Do you leave this for me to implement?

Goldorion commented:
> Do you leave this for me to implement?

If you have more free time than me, I won't say no, but I'll fix them when I will have some free time.

NerdyPuzzle commented:
How does the convertion deal with locked ranged items?

Goldorion commented:
> How does the convertion deal with locked ranged items?

Good question. Surely the same way it handled food mod elements when I deleted it.

Defeatomizer commented:
It most likely ignores the <kbd>locked</kbd> flag state.

KlemenDEV commented:
Please merge master into this PR

Defeatomizer commented on src/main/java/net/mcreator/element/converter/ConverterRegistry.java:
```diff
@@ -79,8 +79,7 @@ public class ConverterRegistry {
 				new LegacyShootArrowProcedureRemover(), new BiomeDictionaryProcedureConverter(),
-				new ToolToItemTypeProcedureConverter()));
-		put(ModElementType.RANGEDITEM, List.of(new RangedItemTextureConverter()));
+				new ToolToItemTypeProcedureConverter(), new ProcedureArrowProjectileFixer()));
```
```suggestion
				new ToolToItemTypeProcedureConverter(), new ExplodeProcedureConverter(),
				new ProcedureArrowProjectileFixer()));
```

Goldorion replied:
I will do it when merging master to avoid any problem

Defeatomizer replied:
No problem, I just showed the spot of the very conflict.

KlemenDEV commented:
This is a message that applies to all MC-code-related PRs. 1.18.2 generator was removed so in case this PR did not do this already, two actions are needed:

* 1.18.2 generator support needs to be removed
* 1.19.4 generator support needs to be added

If those changes were already applied, this message can be ignored.

Goldorion commented on src/main/java/net/mcreator/element/converter/v2023_1/RangedItemToProjectileAndItemConverter.java:
```diff
@@ -0,0 +1,158 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class RangedItemToProjectileAndItemConverter implements IConverter {
```
Can I have the workspace you used? My test workspace (that includes several cases) doesn't have this problem and converts perfectly without Freemarker errors in the console.

KlemenDEV replied:
In fact, this happens with new workspaces too if the custom model is used. To reproduce, just make any projectile that uses a custom model. This makes me wonder how much the whole PR was tested.

Please before marking as ready for review and after you do all changes, do another in-depth test of all features. If needed, go through UI and make sure you have covered all scenarios, as it really does not help that I need to manually review and doubt every component of the PR to make sure things like this do not happen.

KlemenDEV replied:
Here is a workspace made in a few minutes: [projectilemodel.zip](https://github.com/MCreator/MCreator/files/11558389/projectilemodel.zip)

KlemenDEV commented on plugins/generator-1.19.2/forge-1.19.2/templates/item/item.java.ftl:
```diff
@@ -252,5 +255,97 @@ public class ${name}Item extends Item {
+	</#if>
+
+	<#if data.enableRanged && data.shootConstantly>
+		@Override public void onUsingTick(ItemStack itemstack, LivingEntity entity, int count) {
```
This method is deprecated

Goldorion replied:
Can you tell me where you saw the method is deprecated? The code does not say it's deprecated

![image](https://github.com/MCreator/MCreator/assets/38427877/02e2d744-5af4-4b65-8c29-1667c18f5996)

KlemenDEV replied:
This looks deprecated to me:

![image](https://github.com/MCreator/MCreator/assets/16374228/596a8617-26f8-4436-a622-12bc88d31c94)

Please when working on new features, run tests locally and inspect the logs they provide as testing is not as simple as a "green checkmark" on GH PR, all good

Goldorion replied:
If this method is deprecated, I wonder what is the good method because Forge wrote notbing about it in its code (and this is why I used this method)...

Goldorion replied:
Ok, I found the problem. It was Forge, but you with your review. The method is only marked as deprecated since 1.19.4, but you wrote your comment in the 1.19.2 generator, so I thought both versions had this method deprecated (and I checked the code in 1.19.2), while it's only since Forge 1.19.4.

KlemenDEV replied:
Sorry, my bad

KlemenDEV commented on src/main/java/net/mcreator/element/converter/v2023_1/EntitiesRangedAttackConverter.java:
```diff
@@ -17,34 +17,26 @@
  */
 
-package net.mcreator.element.converter.v2022_2;
+package net.mcreator.element.converter.v2023_1;
```
This also needs to be updated on the converters

Goldorion replied:
Should I change it directly to 2023_3 as it will certainly won't be merged before the 3rd snapshot? (And even if it's ready before, I doubt you'll take the risk of merging a big PR just before the full release)

KlemenDEV replied:
Indeed this will not be 2023_2 so you can make it _3 for sure, I just noted this so it is not forgotten

Goldorion replied:
> I just noted this so it is not forgotten

And it was forgotten 😅😬

KlemenDEV commented on src/main/java/net/mcreator/element/converter/v2023_1/EntitiesRangedAttackConverter.java:
```diff
@@ -17,34 +17,26 @@
+import net.mcreator.element.types.LivingEntity;
 import net.mcreator.workspace.Workspace;
 
-public class RangedItemTextureConverter implements IConverter {
```
Nevermind, textures are handled in RangedItemToProjectileAndItemConverter

Goldorion commented on src/main/java/net/mcreator/element/converter/v2023_3/RangedItemToProjectileAndItemConverter.java:
Leaving a review here, so I don't forget.
The test workspace provided in the PR's description seems to no longer work and causes multiple errors in the logs.

Goldorion replied:
I don't know what happened, but it works perfectly now, so I guess I simply did something wrong 🤷

Defeatomizer commented on src/main/java/net/mcreator/element/converter/v2023_1/EntitiesRangedAttackConverter.java:
```diff
@@ -17,34 +17,26 @@
 
 	@Override public int getVersionConvertingTo() {
-		return 31;
+		return 39;
```
FV needs updating here.

Defeatomizer commented on plugins/generator-1.19.2/forge-1.19.2/mappings/projectiles.yaml:
```diff
@@ -1,40 +1,53 @@
 _default:
   - Snowball
   - EntityType.SNOWBALL
+  - Items.AIR
```
Shouldn't most entries here actually map to an item? Snowballs, eggs and others do have their item representation, except for 3 bottom-most projectiles.

Goldorion replied:
Yes, they have. However, the current code of projectiles is only made for custom arrows, while snowballs/eggs and others like fireballs are other types of projectiles (throwable, fireballs, etc.). I already plan to add support for them, but I'm just waiting for this PR to be merged as it would make the PR way too big and it would combine too many things.

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -1115,6 +1115,12 @@ private static GeneratableElement getExampleFor(ModElement modElement, boolean u
+			item.enableRanged = _true;
+			item.projectile = new ProjectileEntry(modElement.getWorkspace(),
+					getRandomDataListEntry(random, ElementUtil.loadArrowProjectiles(modElement.getWorkspace())));
+			item.shootConstantly = _true;
```
You use same condition to enable ranged and to shoot constantly, meaning it will always result in testing ranged item that shoots constantly. I would use eg. emptyLists here to make sure both combinations are tested

KlemenDEV commented on src/main/java/net/mcreator/element/types/Projectile.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Collection;
+import java.util.Collections;
+
+public class Projectile extends GeneratableElement implements IEntityWithModel, ICommonType {
```
This ME could have preview picture that is either texture of projectileItem or if customModel is selected, customModelTexture

KlemenDEV commented:
How is the different model (JSON) for ranged item now handled?

KlemenDEV commented:
Hi! This is a heads-up message that this PR may need some changes related to the recent changes to the MCreator. Keep in mind this is a template message so all of the changes required may not apply to this PR:

* Merge master branch in the PR branch
* Remove 1.19.2 support
* Introduce and test 1.20.1 support
* Remove all help tips translations changes/additions/deletions except for the default/English ones as they are now handled on Crowdin (https://translate.mcreator.net/) the same way as texts.properties

Goldorion commented:
> How is the different model (JSON) for ranged item now handled?

I added a new builtin model, so users can use the ranged item's model (but if the user enables the ranged item panel and the current model is still `Normal`, it will be changed to `Ranged item`)

KlemenDEV commented:
Is this PR ready for review or is it still meant to be draft?

Goldorion commented:
> Is this PR ready for review or is it still meant to be draft?

There is a big problem currently with 1.20.1 I need to fix, but I don'tknow what's causing it. Then, I will have some more testing to do.

KlemenDEV commented:
In this case, I would prefer if you prioritize this before new PRs so we don't clutter the PR backlog, if possible

Goldorion commented:
> Btw if you explain the problem, maybe someone from the community could help you out

To summarize, when I create 2 basic ranged items (items with basic ranged properties), if I add no additional condition it works fine. However, if I add an additional condition, it is not taken in consideration and when I tried to do the condition (tried with if in water), it will shoot constantly.

KlemenDEV commented:
But is is runtime or generated code problem?

Goldorion commented:
Ok, I'm just stupid... I simply forgot I added a stupid procedure (shooting the custom projectile) to multiple triggers (one of the triggers was the projectile entity tick trigger)... so no bug happens, sorry about that... I completely forgot I had those triggers 😅

KlemenDEV commented on src/main/java/net/mcreator/element/converter/v2023_4/RangedItemToProjectileAndItemConverter.java:
```diff
@@ -0,0 +1,179 @@
+/*
```
Two headers

KlemenDEV commented on src/main/java/net/mcreator/element/converter/v2023_4/ProcedureArrowProjectileFixer.java:
```diff
@@ -0,0 +1,107 @@
+/*
```
Two headers, may be more

KlemenDEV commented on plugins/generator-1.20.1/forge-1.20.1/mappings/projectiles.yaml:
```diff
@@ -1,40 +1,52 @@
 Egg:
   - ThrownEgg
   - EntityType.EGG
+  - Items.AIR
```
Shouldn't there be an egg item?

Goldorion replied:
https://github.com/MCreator/MCreator/pull/3456#discussion_r1252240648

KlemenDEV replied:
I think the list should be correct even now though, I think it is better to make it correct now than keep it with air which is a bit confusing until things from the linked comment are implemented.

Or is there a downside?

Goldorion replied:
> Or is there a downside?

The system (that is (should be if I made everything correctly) compatible for for the future) will simply break and might cause more problems. I wanted to make a system already capable, or as most as possible, of supporting more types of projectile in the future without having to change everything again. Using air was the easiest way I found to avoid making even more changes.

KlemenDEV replied:
What will break in this PR if you already set egg to Items.EGG now?

KlemenDEV commented on plugins/generator-1.20.1/forge-1.20.1/mappings/projectiles.yaml:
```diff
@@ -1,40 +1,52 @@
+  - Items.AIR
 ExperienceBottle:
   - ThrownExperienceBottle
   - EntityType.EXPERIENCE_BOTTLE
```
This one lacks 3rd item

Goldorion replied:
https://github.com/MCreator/MCreator/pull/3456#discussion_r1252240648

KlemenDEV reviewed (changes requested):
Please also remove unused key such as "modelement.rangeditem.description" and similar.

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1508,7 +1509,8 @@ modelement.plant.description=Plant is a block that spawns on grass and can be pi
 modelement.potion.description=Potion item is an item giving effects to the entity.
 modelement.procedure.description=Procedure is a graphically designed flow of events with optional global triggers.
-modelement.rangeditem.description=Ranged item shoots entities that act like projectiles and can have custom models.
+modelement.projectile.description=An item that can be shot as an entity using a ranged item.
```
It can be shot in different means too.

I would say something like "Projectile is an entity that can be shot in a direction"

Goldorion commented:
(I'm working on fixing the converter)

KlemenDEV commented:
I have also noticed that eg. when importing RaolCraft, it is not building out of the box after conversion

Goldorion commented:
What's an RC?

Defeatomizer commented:
If the workspace is big, maybe some elements forget to get converted, e.g. if a procedure is updated after element using it (to be resolvable via usages dialog?).

Goldorion commented:
The converted ranged item getting its new item mod element is causing the problem for some reason, but I don't understand why as I didn't touch on this 🤔


```
2023-09-23-09:40:25 [ME preloader/ERROR] [ModElementManager] GeneratableElement cache contains element with same name but different object. This should not happen!
2023-09-23-09:40:25 [ME preloader/DEBUG] [Converter utils] Converted mod element Qqq (rangeditem) to item using RangedItemToProjectileAndItemConverter```

KlemenDEV commented:
> If the workspace is big, maybe some elements forget to get converted, e.g. if a procedure is updated after element using it

References need to be updated by the converter when needed. I can take a look into this part after all other parts of the PR look fine.

Goldorion commented:
> If the workspace is big, maybe some elements forget to get converted, e.g. if a procedure is updated after element using it (to be resolvable via usages dialog?).

Nah, the test workspace I use is 1 procedure and 1 ranged item (I wanted to test the changes I made to the converter with the additions of the special information PR)

KlemenDEV commented:
> The converted ranged item getting its new item mod element is causing the problem for some reason, but I don't understand why as I didn't touch on this 🤔

Could be if new ME has same name as old ME, try adding prefix to both

Goldorion commented:
> Could be if new ME has same name as old ME, try adding prefix to both

I won't say it doesn't work, but all mod elements are now not visible inside the UI (but their JSON files still exist) 😅🤔

Goldorion commented:
> > The converted ranged item getting its new item mod element is causing the problem for some reason, but I don't understand why as I didn't touch on this 🤔
> 
> Could be if new ME has same name as old ME, try adding prefix to both

The same problem still happens even after multiple attempts and different things tried.

Defeatomizer commented:
If an element is called `UncertainMagicProjectile` and before it JSON provides an `UncertainMagic` ranged item to convert, what are your chances to succeed? :thinking:

Goldorion commented:
> If an element is called `UncertainMagicProjectile` and before it JSON provides an `UncertainMagic` ranged item to convert, what are your chances to succeed? 🤔

Good question 🤔 I never thought of this case (and we never tested this case with the FoodToItem and FuelToItemExtension converters)

Goldorion commented:
However, I can't test what happens for the moment because MCreator decided to be stubborn and not work when the converter is applied until the workspace is re-opened 🤷

KlemenDEV commented:
I have made some changes and fixed some bugs, but I see there is still some work to do and more importantly, testing from the PR author.

One thing, for example, is that no JSON is generated for the ranged item model, at least in 1.20.1 as I did not proceed with testing after that. Please, do extensive testing of this PR again as it is quite big, fix any outstanding bugs, and then mark it as ready for review.

KlemenDEV commented:
It would be good if this PR could be ready for the next snapshot, otherwise, it will probably need another porting to 1.20.2 as 2024.1 will very likely update MC version

KlemenDEV commented:
When you are done with testing, mark for review so I know I can start reviewing. Thanks!

Goldorion commented:
(Re) marking as a draft because I just noticed a problem with one of my test item in 1.19.4. so I'll investigate. However, 1.20.1 works fine, so it is ready for review.

KlemenDEV commented on src/main/java/net/mcreator/element/converter/v2023_4/RangedItemToProjectileAndItemConverter.java:
```diff
@@ -0,0 +1,182 @@
+			else
+				item.animation = "bow";
+
+			if (rangedItem.has("glowCondition"))
```
You are checking for glowCondition here and then again in else if, there is something wrong here likely

Goldorion replied:
Oh, I forgot to remove it when changing the converter (to apply the glow feature converter you made)... but it never caused a problem during my tests 🤷

KlemenDEV replied:
I don't think it will cause problems, but it is still unclean code

Goldorion replied:
It didn't seem to cause a problem during my tests, so this is why I didn't notice my mistake 🫠

Goldorion replied:
I re-tested and this still works fine.

KlemenDEV reviewed (changes requested):
Old ranged items shot with power depending on the "charge time" but now this is no longer the case. Please make it act on charge time again

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/templates/projectile/projectile.java.ftl:
```diff
@@ -166,21 +162,18 @@ public class ${name}Entity extends AbstractArrow implements ItemSupplier {
 		</#if>
 		entity.level.addFreshEntity(entityarrow);
-
-		<#if data.actionSound.toString()?has_content>
```
Before, action sound was optional, and this should stay so (no call to this if not needed)

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/templates/projectile/projectile.java.ftl:
```diff
@@ -141,22 +135,24 @@ public class ${name}Entity extends AbstractArrow implements ItemSupplier {
 		</#if>
 		world.addFreshEntity(entityarrow);
 
-		<#if data.actionSound.toString()?has_content>
```
Before, action sound was optional, and this should stay so (no call to this if not needed)

KlemenDEV reviewed (changes requested):
Applies to both versions

Goldorion commented:
> Old ranged items shot with power depending on the "charge time" but now this is no longer the case. Please make it act on charge time again

I just did some tests and the behaviour is exactly the same as in MCreator 2023.3. In 2023.3, I made a ranged item and tried to kill an iron golem with no charging state and then I tried again with the item fully charged and it took the same amount of projectiles before killing the iron golem. If you do the same test with this snapshot, both ways will require the same amount of projectiles.

If you can't charge, just check your "Item usage duration". If it's 0, you need to change it for a bigger value (default is 72 000)

KlemenDEV reviewed (changes requested):
The converter will need more work. I have just made ranged item in master branch and then switched to this branch and there is this error:

https://pastebin.com/twLrnXDP

![image](https://github.com/MCreator/MCreator/assets/16374228/7f4dcf12-4878-4c82-b28f-bd6721d6b841)

Please, do more testing. I can't doubt and test every single feature of the PR (although currently I do), it is not time effective

Goldorion commented:
I tested using the test workspace attached to the PR and I also created a new workspace using 2023.3, but I admit I forgot to test using the new 2023.4 format. It is now fixed (and re-tested with the test workspace and master branch)

KlemenDEV commented on plugins/generator-1.19.4/forge-1.19.4/templates/projectile/projectile.java.ftl:
```diff
@@ -38,6 +38,8 @@ package ${package}.entity;
 @OnlyIn(value = Dist.CLIENT, _interface = ItemSupplier.class)
 public class ${name}Entity extends AbstractArrow implements ItemSupplier {
 
+	public static final ItemStack PROJECTILE_ITEM = ${mappedMCItemToItem(data.projectileItem)}.getDefaultInstance();
```
Why is getDefaultInstance call needed instead of mappedMCItemToItemStackCode how it was done before and is done in other locations where itemstack is needed?

Goldorion replied:
It is needed in the item mod element (arrowshootcode macro if I remember correctly)

KlemenDEV commented on src/main/java/net/mcreator/element/types/Item.java:
```diff
@@ -91,6 +92,13 @@
+	public boolean shootConstantly;
+	public ProjectileEntry projectile;
+	public Procedure onRangedItemUsed;
+	public Procedure useCondition;
```
useCondition is too loose term as this explicitly refers to ranged use condition and not to general use condition. Please refactor all related to this, including translation and help tip key

Defeatomizer replied:
`shootCondition` would be a better name?

Goldorion replied:
I was thinking to `useRangedCondition`

KlemenDEV replied:
Or even better rangedUseCondition to follow how it is called in Dimension for portal

KlemenDEV replied:
Thanks for the quick fix!

KlemenDEV commented:
RFR?

Goldorion commented:
Sorry, I accidently clicked on the button.

Defeatomizer commented on src/main/java/net/mcreator/element/converter/v2023_4/RangedItemToProjectileAndItemConverter.java:
```diff
@@ -116,10 +116,13 @@ else if (rangedItem.get("ammoItem") != null && !rangedItem.get("ammoItem").getAs
+				if (item.customModelName.equals("Normal") && item.renderType == 0)
+					item.customModelName = "Ranged item"; // Convert the ranged items' normal model to items' RI model option
+			} else {
+				item.customModelName = "Ranged item";
```
Should there be another model assigned?

KlemenDEV replied:
If nothing is defined, ranged item should be assumed

Goldorion replied:
So is it fine or not?

KlemenDEV replied:
I think it is fine

Defeatomizer replied:
My bad, in pull request diff it looked like these two assignments were on the same level.

Goldorion commented:
It's finally merged... after so much time (ok, less than a year, but still). Projectiles will finally get some love with lots of new features 👀
